pkg/parser/instance/proto.go

```diff
@@ -0,0 +1,141 @@
+package instance
+
+import (
+	"github.com/cirruslabs/cirrus-cli/pkg/parser/nameable"
+	"github.com/cirruslabs/cirrus-cli/pkg/parser/node"
+	"github.com/cirruslabs/cirrus-cli/pkg/parser/parseable"
+	"github.com/cirruslabs/cirrus-cli/pkg/parser/schema"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+	"strconv"
+	"strings"
+)
+
+type ProtoInstance struct {
+	proto *dynamicpb.Message
+
+	parseable.DefaultParser
+}
+
+//nolint:gocognit // it's a parser, there is a lot of boilerplate
+func NewProtoParser(desc protoreflect.MessageDescriptor, mergedEnv map[string]string) *ProtoInstance {
+	instance := &ProtoInstance{
+		proto: dynamicpb.NewMessage(desc),
+	}
+
+	fields := desc.Fields()
+	for i := 0; i < fields.Len(); i++ {
+		field := fields.Get(i)
+		switch field.Kind() {
+		case protoreflect.MessageKind:
+			instance.OptionalField(nameable.NewSimpleNameable(field.JSONName()), schema.TodoSchema, func(node *node.Node) error {
+				childParser := NewProtoParser(field.Message(), mergedEnv)
+				parserChild, err := childParser.Parse(node)
+				if err != nil {
+					return err
+				}
+				instance.proto.Set(field, protoreflect.ValueOfMessage(parserChild))
+				return nil
+			})
+		case protoreflect.EnumKind:
+			instance.OptionalField(nameable.NewSimpleNameable(field.JSONName()), schema.TodoSchema, func(node *node.Node) error {
+				value, err := node.GetExpandedStringValue(mergedEnv)
+				if err != nil {
+					return err
+				}
+				enumValueDescriptor := field.Enum().Values().ByName(protoreflect.Name(strings.ToUpper(value)))
+				instance.proto.Set(field, protoreflect.ValueOfEnum(enumValueDescriptor.Number()))
+				return nil
+			})
+		case protoreflect.StringKind:
+			instance.OptionalField(nameable.NewSimpleNameable(field.JSONName()), schema.TodoSchema, func(node *node.Node) error {
+				value, err := node.GetExpandedStringValue(mergedEnv)
+				if err != nil {
+					return err
+				}
+				instance.proto.Set(field, protoreflect.ValueOfString(value))
+				return nil
+			})
+		case protoreflect.Int64Kind, protoreflect.Sint64Kind, protoreflect.Uint64Kind,
+			protoreflect.Fixed64Kind, protoreflect.Sfixed64Kind:
+			instance.OptionalField(nameable.NewSimpleNameable(field.JSONName()), schema.TodoSchema, func(node *node.Node) error {
+				value, err := node.GetExpandedStringValue(mergedEnv)
+				if err != nil {
+					return err
+				}
+				parsedValue, err := strconv.ParseInt(value, 10, 64)
```

Comment:
Shouldn't that be in `else` branch of the following `if`?

Reply:
See c47a3ef